### ISSUE TITLE
using intersection observer api

### ISSUE THREAD

**Issue by Mohsen-Khakbiz** [jlmakes/scrollreveal#381]
is there anyway to use intersection observer api for this plugin? you  know for better performance.
as there's a polyfill for IO there's no concern about browser compatibility in my case. 
or if you could guide me how to achieve it, it'll be much appreciated.

Regards,
Mohsen

**Comment by jlmakes** [comment]
Thanks for asking Mohsen.

## Performance

I’m not aware of any performance issues with ScrollReveal; in fact, the only issue I can recall was related to [layout reflow](#335), and it was improved significantly in [ScrollReveal 4 (beta)](https://github.com/jlmakes/scrollreveal/tree/development).

Other improvements include caching all element dimensions and position, further lightening the burden on event handlers. Even though it’s in beta (for only a few more weeks), version 4 is the most stable and robust version to date—I strongly recommend upgrading.

```html
<script src="https://cdn.jsdelivr.net/npm/scrollreveal@beta"></script>
```

See [What’s New in 4.0](https://github.com/jlmakes/scrollreveal/blob/development/CHANGELOG.md).

## Intersection Observer

I’d like to detail, is that Intersection Observer will include CSS Transforms in its return values—which actually complicates using them for reveal on scroll libraries, particularly ScrollReveal. Let’s look at an example of how this makes our lifer harder.

> I know `IntersectionObserver` doesn’t exactly use coordinations (but rather percentages of visibility), but the idea behind this example still applies. Please bear with me.

Imagine we have an Element at coordinates (2, 1) on our page:

![screen shot 2017-10-11 at 4 52 53 pm](https://user-images.githubusercontent.com/2044842/31448422-b3ba9a06-aea4-11e7-8b01-cfdd0750dce8.png)

ScrollReveal (and I presume other reveal on scroll libraries) will use CSS Transform to create animations. Let’s say we’ve added some `translateX` to our Element because we’re planning to reveal it from the right side of the window.

![screen shot 2017-10-11 at 4 53 01 pm](https://user-images.githubusercontent.com/2044842/31448453-c5f522ae-aea4-11e7-9d03-96b5a3344eb3.png)

What would we expect/want the coordinates of our Element to be?

This is where we see the critical difference between using `offsetLeft` and `offsetTop` to determine Element visibility, and using modern browser APIs such as `IntersectionObserver` or `getBoundingClientRect`.

If we use the newer APIs, they will _include_ the CSS Transform, we we’d get (9, 1):

![screen shot 2017-10-11 at 5 02 12 pm](https://user-images.githubusercontent.com/2044842/31449026-29b6733c-aea6-11e7-8e76-99826cf28573.png)

Whereas, if we use offset values, we’ll still get the original coordinates (2, 1)!

![screen shot 2017-10-11 at 5 07 43 pm](https://user-images.githubusercontent.com/2044842/31449258-b8664b8e-aea6-11e7-8940-891c19a3f117.png)

### Why This Matters

Well, is the Element (9, 1) within the window's range of {8, 5}?

No. It’s not within the window (which is easily seen in the figure). In fact, this element would never be considered "visible" (without some form of horizontal scrolling), but herein lies the problem.

**We don’t really care where the element is in its pre-reveal state... rather, we want to know "if we revealed this element now, would it be visible after the animation completes?"**

It _sounds_ like a good thing that modern APIs include our CSS Transform, and indeed for many use cases I’m sure it is... but by obscuring the element’s original/post-reveal position, it actually makes it significantly more difficult to determine whether our Element should be revealed.

> I actually tried to use `getBoundingClientRect` at one point, and elements would constantly bounce in and out of view because the position coordinates were constantly changing during animation.

Now, `IntersectionObserver` uses percentages (something called Intersection Ratios), not coordinates, but the problem is the same: The position reported by `IntersectionObserver` is influenced by CSS Transform.

> In fact, I actually observed some odd behavior combining CSS Transform with `IntersectionObserver`, but I didn’t explore it thoroughly.

### So, Offsets Forever?

Well, because offset values ignore CSS transforms, we can trust that when we check an element’s position, we will be checking the "post-reveal" position we care about... not the rendered position of the element.

I’m sure a clever developer could build a reveal on scroll library from scratch that leverages `getBoundingClientRect` or `IntersectionObserver`, but ScrollReveal has a pretty substantial feature base with very wide browser support.

## tl;dr

I wouldn’t recommend trying to replace the element visibility logic with `IntersectionObserver` ...or `getBoundingClientRect` for that matter.

**Comment by Mohsen-Khakbiz** [comment]
wow, thanks for the detailed information. appreciate it. learned somethings I didn't now.
version 4 is doing a pretty much good job now. I was looking for a solution to make this landing page work smoother http://boo.themerella.com/boo-landing/
but after adding the latest ScrollReveal beta version, it significantly improved the performance.

thanks for the awesome plugin
